### DISCUSSION TITLE
routing behind reverse proxies (use x-forwarded-host for localized routing)

### DISCUSSION THREAD

**Discussion by gijsbotje** [vercel/next.js#34571] [Ideas]
### Describe the feature you'd like to request

Our stack runs a next site in a docker container. We would like to be able to use domain routing for localization. Unfortunately, the `host` header is the URL of the container it's deployed on. We figured out that the `x-forwarded-host` wasn't set, so we did in our proxy. When I dug deeper in the code of how next and the i18n module does the routing, I found that it only looks at the `host` header.

### Describe the solution you'd like

I found https://github.com/vercel/next.js/pull/20662 which adds a config option to trust proxies. This would fix our problems with the headers. 
Or if anyone has a workaround to also include the `x-forwarded-host` in the `getHostname` functions, that could also help.

### Describe alternatives you've considered

I haven't found any other alternative for our particular deployment.
https://github.com/vercel/next.js/pull/20662 would be perfect and I would be happy to make a new PR that also ads this to the i18n module.

**Comment by dlindenkreuz** [comment]
I have the same problem. I would expect a `trustProxy` setting that is not tied to i18n, but instead works on a global level: It should affect all `request` objects in middleware and API routes.

When running behind a reverse proxy like nginx, the `trustProxy` setting should consider all `x-forwarded-{host,port,proto}` headers and modify request URLs accordingly.

This behavior is described and already implemented by [Express.js](https://expressjs.com/en/guide/behind-proxies.html). Maybe Next.js can take some cues from their implementation.

**Next.js v12.0.9 added some urgency to this,** as [absolute URLs are enforced at a middleware/edge function level](https://github.com/vercel/next.js/pull/33410). I now need to somehow generate absolute URLs and will get errors when using relative URLs.

There are two possible workarounds:

1. Use a static `APP_BASE_URL=https://app.domain.xyz` env variable to construct absolute URLs, e.g. for middleware redirects
2. Construct absolute URLs using `x-forwarded-{host,port,proto}` inside `req.headers`

**Comment by dennisroethig** [comment]
Hi, just came across this as I was looking for the same solution. Has there been any progress/thinking on this from the Vercel side @timneutkens? 🙇

**Comment by dennisroethig** [comment]
Hey @timneutkens - could you point me in the right direction to get the conversation started about this one? This is still a big blocker for our setup and I would love to find out how we can make something happen 🙇 😄

**Comment by prokopsimek** [comment]
Hello! I was looking for a solution to the same problem, so I'm adding my current solution for people who come here to this thread. 😊  

```typescript
/* eslint-disable @next/next/no-server-import-in-page */
import { NextRequest, NextResponse } from 'next/server';

export function middleware(request: NextRequest) {
  const url = request.nextUrl.clone();

  const isProduction = process.env.NODE_ENV === 'production' // redirect only in production
  const requestedHost = request.headers.get('X-Forwarded-Host');

  if (isProduction && requestedHost && !requestedHost.match(/example.com/)) {
    const host = `example.com`; // set your main domain

    const requestedPort = request.headers.get('X-Forwarded-Port');
    const requestedProto = request.headers.get('X-Forwarded-Proto');

    url.host = host;
    url.protocol = requestedProto || url.protocol;
    url.port = requestedPort || url.port;

    return NextResponse.redirect(url);
  }

  return NextResponse.next();
}
```
https://stackoverflow.com/a/74991466/10826693
